packages/sdk/src/animation/animationData.ts

```diff
@@ -0,0 +1,226 @@
+	keyframes: Array<Keyframe<T>>;
+	/** Whether the keyframe values are relative to 0 or to the target's current property value. Defaults to false. */
+	relative?: boolean;
+}
+
+/** The value of an animation property at a moment in time */
+export type Keyframe<T extends AnimationProp> = {
+	/** The time in seconds from the start of the animation.  */
+	time: number;
+	/** The property's value at this instant, or a reference to another property. */
+	value: Like<T> | TargetPath<T>;
+	/** How the value approaches this frame's value. Defaults to linear. */
+	easing?: EaseCurve;
+}
+
+/** Keyframe data for an animation */
+export interface AnimationDataLike {
+	/** The animation keyframe data */
+	tracks: Readonly<Array<Track<AnimationProp>>>;
+}
+
+export class AnimationData extends Asset implements AnimationDataLike, Patchable<AssetLike> {
+	private _internal = new AssetInternal(this);
+	public get internal() { return this._internal; }
+
+	private _tracks: Array<Track<AnimationProp>>;
+
+	/** @inheritdoc */
+	public get tracks() { return this._tracks; }
+
+	/** @inheritdoc */
+	public get animationData(): AnimationData { return this; }
+
+	/** @hidden */
+	public constructor(container: AssetContainer, def: AssetLike) {
+		super(container, def);
+
+		if (!def.animationData) {
+			throw new Error("Cannot construct mesh from non-animation data definition");
+		}
+
+		this.copy(def);
+	}
+
+	/** The length of the animation data. */
+	public duration() {
+		return this.tracks.reduce((trackMax, track) =>
+			Math.max(trackMax, track.keyframes.reduce((kfMax, kf) =>
+				Math.max(kfMax, kf.time), 0)), 0);
+	}
+
+	/** The placeholders and types of the animation's targets. */
+	public targets() {
+		return this.tracks
+			// get root objects
+			.map(t => t.target.toString().split('/')[0])
+			// that are unique in the list
+			.filter((target, i, arr) => arr.indexOf(target) === i)
+			// and add their types/names to an object
+			.reduce((obj, id) => {
+				const [typeName, placeholder] = id.split(':');
+				if (/^actor|animation|material$/u.test(typeName)) {
+					obj[placeholder] = typeName as AnimatibleName;
+				}
+				return obj;
+			}, {} as {[placeholder: string]: AnimatibleName});
+	}
+
+	/**
+	 * Bind this animation data to one or more targets to create an [[Animation]].
+	 * @param targets A map of placeholder names to real objects. The names and types must match those in [[targets]].
+	 * @param initialState Initial properties for the new animation.
+	 */
+	public bind(
+		targets: { [placeholder: string]: Animatible },
+		initialState?: Partial<AnimationLike>
+	): Promise<Animation> {
+		// validate names and types
+		const dataTargets = this.targets();
+		const dataPlaceholders = new Set(Object.keys(dataTargets));
+		const targetIds: { [placeholder: string]: Guid } = {};
+
+		for (const placeholder in targets) {
+			if (!dataPlaceholders.has(placeholder)) {
+				throw new Error(`Animation data "${this.name || this.id}" has no reference ` +
```

Comment:
You're introducing a lot of new thrown `Error`s in this changeset. Thrown exceptions aren't a great error propagation mechanism in SDKs because they're not encoded into the function signature in a type-aware way. I think we need a way of returning errors that is reflective (for intellisense), and doesn't require try/catch or null checking everywhere. These new `throw`s can stay in this PR/feature, but let's think of them as stand-ins for a better system. I want to commit to never propagating exceptions from the public API.

I'll do some research and suggest a system we can use across all our public APIs. I'm thinking of something like the `Either` struct from fp-ts: https://gcanti.github.io/fp-ts/modules/Either.ts.html. Essentially, return a value that is either an Error or the result. With some helper methods like `onSuccess`, `onError` added. But let me try some experiments to explore it.

I do think it's appropriate to let platform exceptions propagate. If a file open fails or an exception from another API occurs, let those bubble up to the app. This is really about returning the errors we check for in a uniform, friendly way.

Reply:
Try-Catch is a cross-language standard, and integrated with the Async-Await flow. I agree that known exception cases should be documented (maybe we can set up a lint rule), but all these other structures seem unnecessary when try-catch accomplishes the same thing in a supported way.